### ISSUE TITLE
SIGILL: Latest stable Mono crash when loading an XML Document

### ISSUE THREAD

**Issue by sbrl** [mono/mono#13277]
## Steps to Reproduce

1. Load an XML document in with System.Xml (offending code is in [this repository](https://git.starbeamrainbowlabs.com/sbrl/RhinoReminds))
2. 
3. 

### Current Behaviour
It crashes with a native stack trace

### Expected Behaviour
It doesn't crash and actually loads the following XML document:

```
﻿<?xml version="1.0" encoding="utf-8"?>
<ReminderList>
  <NextId>151</NextId>
  <Reminders />
</ReminderList>
```

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**: `Mono JIT compiler version 5.18.0.268 (tarball Thu Feb 28 15:24:08 UTC 2019)` (from `deb http://download.mono-project.com/repo/ubuntu stable-bionic main`)

## Stacktrace

```=================================================================
	Native Crash Reporting
=================================================================
Got a SIGILL while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
/proc/self/maps:
40e46000-40e56000 rwxp 00000000 00:00 0 
41bb2000-41c02000 rwxp 00000000 00:00 0 
5628d0f63000-5628d13af000 r-xp 00000000 08:01 783582                     /usr/bin/mono-sgen
5628d15ae000-5628d15b5000 r--p 0044b000 08:01 783582                     /usr/bin/mono-sgen
5628d15b5000-5628d15ba000 rw-p 00452000 08:01 783582                     /usr/bin/mono-sgen
5628d15ba000-5628d15d1000 rw-p 00000000 00:00 0 
5628d2ed7000-5628d324b000 rw-p 00000000 00:00 0                          [heap]
7f6bc8000000-7f6bc8021000 rw-p 00000000 00:00 0 
7f6bc8021000-7f6bcc000000 ---p 00000000 00:00 0 
7f6bcd700000-7f6bcd800000 rw-p 00000000 00:00 0 
7f6bcd844000-7f6bcdb4c000 r--p 00000000 08:01 783977                     /usr/lib/mono/gac/System.Xml/4.0.0.0__b77a5c561934e089/System.Xml.dll
7f6bcdb4c000-7f6bcddc0000 r--p 00000000 08:01 793357                     /usr/lib/mono/gac/System/4.0.0.0__b77a5c561934e089/System.dll
7f6bcddc0000-7f6bcddc1000 ---p 00000000 00:00 0 
7f6bcddc1000-7f6bcddc2000 rw-p 00000000 00:00 0 
7f6bcddc2000-7f6bcddca000 ---p 00000000 00:00 0 
7f6bcddca000-7f6bcdfc1000 rw-p 00000000 00:00 0 
7f6bcdfc1000-7f6bce57d000 r-xp 00000000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7f6bce57d000-7f6bce77c000 ---p 005bc000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7f6bce77c000-7f6bce77d000 r--p 005bb000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7f6bce77d000-7f6bce77e000 rw-p 005bc000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7f6bce77e000-7f6bce7a8000 rw-p 00000000 00:00 0 
7f6bce7a8000-7f6bcebff000 r--p 00000000 08:01 783516                     /usr/lib/mono/4.5/mscorlib.dll
7f6bcebff000-7f6bcfbff000 rw-p 00000000 00:00 0 
7f6bcfbff000-7f6bcfc00000 ---p 00000000 00:00 0 
7f6bcfc00000-7f6bd0800000 rw-p 00000000 00:00 0 

=================================================================
	Basic Fault Adddress Reporting
=================================================================
Memory around native instruction pointer (0x7f6bce30efb9):0x7f6bce30efa9  20 48 09 ca 41 0f b6 4b 05 48 c1 e1 28 48 09 d1   H..A..K.H..(H..
0x7f6bce30efb9  66 41 0f 38 32 4b 06 66 0f 6f 05 68 d0 26 00 66  fA.82K.f.o.h.&.f
0x7f6bce30efc9  0f 38 00 c8 66 49 0f 7e ca 66 48 0f 3a 16 ca 01  .8..fI.~.fH.:...
0x7f6bce30efd9  4c 09 d2 48 09 ca 45 0f b6 c9 41 0f b6 4b 09 48  L..H..E...A..K.H

=================================================================
	Native stacktrace:
=================================================================
	0x5628d108d9a8 - /usr/bin/mono : (null)
	0x5628d1023284 - /usr/bin/mono : (null)
	0x5628d0fa4618 - /usr/bin/mono : (null)
	0x7f6bd137a890 - /lib/x86_64-linux-gnu/libpthread.so.0 : (null)
	0x7f6bce30efb9 - /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so : (null)

=================================================================
	Telemetry Dumper:
=================================================================
Pkilling 0x7f6bcdfc0700 from 0x7f6bd1f43780
Entering thread summarizer pause from 0x7f6bd1f43780
Finished thread summarizer pause from 0x7f6bd1f43780.

Waiting for dumping threads to resume

Debug info from gdb:


=================================================================
	External Debugger Dump:
=================================================================
mono_gdb_render_native_backtraces not supported on this platform, unable to find gdb or lldb

=================================================================
	Managed Stacktrace:
=================================================================
	  at System.Globalization.CompareInfo:CompareOrdinalIgnoreCase <0x00129>
	  at System.String:Compare <0x000b6>
	  at System.Xml.XmlTextReaderImpl:CheckEncoding <0x0007a>
	  at System.Xml.XmlTextReaderImpl:ParseXmlDeclaration <0x00de7>
	  at System.Xml.XmlTextReaderImpl:Read <0x000bf>
	  at System.Xml.XmlLoader:Load <0x001f0>
	  at System.Xml.XmlDocument:Load <0x0008f>
	  at System.Xml.XmlDocument:Load <0x0007f>
	  at RhinoReminds.ReminderList:FromXmlFile <0x00078>
	  at <Start>d__27:MoveNext <0x000fb>
	  at System.Runtime.CompilerServices.AsyncTaskMethodBuilder:Start <0x000ac>
	  at RhinoReminds.ClientListener:Start <0x0015a>
	  at RhinoReminds.Program:run <0x00373>
	  at RhinoReminds.Program:Main <0x00997>
	  at <Module>:runtime_invoke_int_object <0x000f4>
=================================================================
```

## Additional details
 - `uname -a`: `Linux starbeamrainbowlabs.com 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
 - OS: `Ubuntu 18.04.2 LTS` (server)
 - CPU: `Intel(R) Atom(TM) CPU N2800   @ 1.86GHz` ([lscpu](https://starbeamrainbowlabs.com/bin/l8xq), available for 60 days)
 -

**Comment by lewurm** [comment]
<img width="641" alt="screenshot 2019-03-04 at 09 14 21" src="https://user-images.githubusercontent.com/75403/53719278-ef595300-3e5d-11e9-8186-3023a112235d.png">

`pmovzxbq` is part of the SSE4.1 instruction set extension and it looks like it isn't supported by your CPU. Sounds like issue https://github.com/mono/mono/issues/12182 and in particular this workaround should get you unblocked: https://github.com/mono/mono/issues/12182#issuecomment-450971306

**Comment by sbrl** [comment]
Hey, @lewurm! Thanks for the reply. I've done this:

```
$ sudo apt purge mono-llvm-support
[sudo] password for sbrl: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  mono-llvm-tools
Use 'sudo apt autoremove' to remove it.
The following packages will be REMOVED
  mono-llvm-support*
0 to upgrade, 0 to newly install, 1 to remove and 13 not to upgrade.
After this operation, 10.8 MB disk space will be freed.
Do you want to continue? [Y/n] 
(Reading database ... 193431 files and directories currently installed.)
Removing mono-llvm-support (5.18.0.268-0xamarin7+ubuntu1804b1) ...
Processing triggers for libc-bin (2.27-3ubuntu1) ...
```

....but I'm still getting this error:

```
=================================================================
	Native Crash Reporting
=================================================================
Got a SIGILL while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
/proc/self/maps:
40a2a000-40a7a000 rwxp 00000000 00:00 0 
410be000-410ce000 rwxp 00000000 00:00 0 
560ca3cd6000-560ca4122000 r-xp 00000000 08:01 783582                     /usr/bin/mono-sgen
560ca4321000-560ca4328000 r--p 0044b000 08:01 783582                     /usr/bin/mono-sgen
560ca4328000-560ca432d000 rw-p 00452000 08:01 783582                     /usr/bin/mono-sgen
560ca432d000-560ca4344000 rw-p 00000000 00:00 0 
560ca5cd3000-560ca6050000 rw-p 00000000 00:00 0                          [heap]
7ff02c000000-7ff02c021000 rw-p 00000000 00:00 0 
7ff02c021000-7ff030000000 ---p 00000000 00:00 0 
7ff030700000-7ff030800000 rw-p 00000000 00:00 0 
7ff030844000-7ff030b4c000 r--p 00000000 08:01 783977                     /usr/lib/mono/gac/System.Xml/4.0.0.0__b77a5c561934e089/System.Xml.dll
7ff030b4c000-7ff030dc0000 r--p 00000000 08:01 793357                     /usr/lib/mono/gac/System/4.0.0.0__b77a5c561934e089/System.dll
7ff030dc0000-7ff030dc1000 ---p 00000000 00:00 0 
7ff030dc1000-7ff030dc2000 rw-p 00000000 00:00 0 
7ff030dc2000-7ff030dca000 ---p 00000000 00:00 0 
7ff030dca000-7ff030fc1000 rw-p 00000000 00:00 0 
7ff030fc1000-7ff03157d000 r-xp 00000000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7ff03157d000-7ff03177c000 ---p 005bc000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7ff03177c000-7ff03177d000 r--p 005bb000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7ff03177d000-7ff03177e000 rw-p 005bc000 08:01 783539                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7ff03177e000-7ff0317a8000 rw-p 00000000 00:00 0 
7ff0317a8000-7ff031bff000 r--p 00000000 08:01 783516                     /usr/lib/mono/4.5/mscorlib.dll
7ff031bff000-7ff032bff000 rw-p 00000000 00:00 0 
7ff032bff000-7ff032c00000 ---p 00000000 00:00 0 
7ff032c00000-7ff033800000 rw-p 00000000 00:00 0 

=================================================================
	Basic Fault Adddress Reporting
=================================================================
Memory around native instruction pointer (0x7ff03130efb9):0x7ff03130efa9  20 48 09 ca 41 0f b6 4b 05 48 c1 e1 28 48 09 d1   H..A..K.H..(H..
0x7ff03130efb9  66 41 0f 38 32 4b 06 66 0f 6f 05 68 d0 26 00 66  fA.82K.f.o.h.&.f
0x7ff03130efc9  0f 38 00 c8 66 49 0f 7e ca 66 48 0f 3a 16 ca 01  .8..fI.~.fH.:...
0x7ff03130efd9  4c 09 d2 48 09 ca 45 0f b6 c9 41 0f b6 4b 09 48  L..H..E...A..K.H

=================================================================
	Native stacktrace:
=================================================================
	0x560ca3e009a8 - /usr/bin/mono : (null)
	0x560ca3d96284 - /usr/bin/mono : (null)
	0x560ca3d17618 - /usr/bin/mono : (null)
	0x7ff03429a890 - /lib/x86_64-linux-gnu/libpthread.so.0 : (null)
	0x7ff03130efb9 - /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so : (null)

=================================================================
	Telemetry Dumper:
=================================================================
Pkilling 0x7ff030fc0700 from 0x7ff034e63780
Entering thread summarizer pause from 0x7ff034e63780
Finished thread summarizer pause from 0x7ff034e63780.

Waiting for dumping threads to resume

Debug info from gdb:


=================================================================
	External Debugger Dump:
=================================================================
mono_gdb_render_native_backtraces not supported on this platform, unable to find gdb or lldb

=================================================================
	Managed Stacktrace:
=================================================================
	  at System.Globalization.CompareInfo:CompareOrdinalIgnoreCase <0x00129>
	  at System.String:Compare <0x000b6>
	  at System.Xml.XmlTextReaderImpl:CheckEncoding <0x0007a>
	  at System.Xml.XmlTextReaderImpl:ParseXmlDeclaration <0x00de7>
	  at System.Xml.XmlTextReaderImpl:Read <0x000bf>
	  at System.Xml.XmlLoader:Load <0x001f0>
	  at System.Xml.XmlDocument:Load <0x0008f>
	  at System.Xml.XmlDocument:Load <0x0007f>
	  at RhinoReminds.ReminderList:FromXmlFile <0x00078>
	  at <Start>d__27:MoveNext <0x000fb>
	  at System.Runtime.CompilerServices.AsyncTaskMethodBuilder:Start <0x000ac>
	  at RhinoReminds.ClientListener:Start <0x0015a>
	  at RhinoReminds.Program:run <0x00373>
	  at RhinoReminds.Program:Main <0x00997>
	  at <Module>:runtime_invoke_int_object <0x000f4>
=================================================================
```

It sounds like the full output of `lscpu` is important, so I'll include that below.

<details>
<summary>Click to see output of <code>lscpu</code></summary>
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  2
Core(s) per socket:  2
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               54
Model name:          Intel(R) Atom(TM) CPU N2800   @ 1.86GHz
Stepping:            1
CPU MHz:             799.836
CPU max MHz:         1862.0000
CPU min MHz:         798.0000
BogoMIPS:            3733.23
L1d cache:           24K
L1i cache:           32K
L2 cache:            512K
NUMA node0 CPU(s):   0-3
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx lm constant_tsc arch_perfmon pebs bts nopl nonstop_tsc cpuid aperfmperf pni dtes64 monitor ds_cpl est tm2 ssse3 cx16 xtpr pdcm movbe lahf_lm dtherm arat
```
</details>

**Comment by lewurm** [comment]
You also need to clear the AOT cache: https://github.com/mono/mono/issues/12182#issuecomment-451160265

If that should help, try this: https://github.com/mono/mono/issues/12182#issuecomment-451118994

**Comment by sbrl** [comment]
Hey, thanks! Clearing the AOT cache does appear to have resolved the issue for now.

If I understand this right, will mono repopulate this cache automatically for me? Is there a command I can use to force it to do so, like it does when install mono (which takes ages, which is why I'm somewhat concerned)? Or does the workaround disable this functionality?

**Comment by marek-safar** [comment]
@lewurm what is the correct solution for the issue on our side?

**Comment by lewurm** [comment]
@sbrl `dpkg-reconfigure libmono-corlib4.5-cil` shoud recreate the AOT cache. Make sure `mono-llvm-support` isn't installed, so it won't use LLVM (which triggers your issue). For what it's worth, everything should work fine without the AOT cache.

@marek-safar we need to fix LLVM AOT for amd64, as discussed in https://github.com/mono/mono/issues/12182

I'll close this issue since it's a duplicate of https://github.com/mono/mono/issues/12182

**Comment by sbrl** [comment]
Thanks, @lewurm! I'll subscribe to that issue then.
